Public/Src/Pips/Dll/Operations/Process.cs

```diff
@@ -765,6 +793,119 @@ public bool TryComputePipUniqueOutputHash(PathTable pathTable, out long pipUniqu
             return true;
         }
 
+        /// <summary>
+        /// Compute the intersection of the pip's dependencies and global affected SourceChangeAffectedOutputs of the build
+        /// </summary>
+        public ReadOnlyArray<string> GetChangeAffectedInputNames(PathTable pathTable)
+        {
+            var changeAffectedInputs = SourceChangeAffectedOutputFiles.Intersect(Dependencies.Select(d => d.Path)).Select(i => i.GetName(pathTable).ToString(pathTable.StringTable)).ToHashSet();
+
+            foreach (var file in SourceChangeAffectedOutputFiles)
+            {
+                foreach (var directory in DirectoryDependencies)
+                {
+                    if (directory.Path.TryGetRelative(pathTable, file, out var relativePath))

```

Comment:
we have a helper that just performs the check called `isWithin` which should be faster since it doesn't have to allocate and compute the result